lib/agent0/agent0/hyperdrive/policies/lpandarb.py

```diff
@@ -0,0 +1,660 @@
+    )
+    y_term = (bond_reserves + bonds_in) ** (FixedPoint(1) - time_stretch)
+    z_val = (k_t - y_term) / (share_price / initial_share_price)
+    z_val = z_val ** (FixedPoint(1) / (FixedPoint(1) - time_stretch))
+    z_val /= initial_share_price
+    # z_val *= one_block_return  # unclear if this helps improve accuracy
+    spot_price = calc_spot_price_local(initial_share_price, share_reserves, FixedPoint(0), bond_reserves, time_stretch)
+    price_discount = FixedPoint(1) - spot_price
+    amount_in_shares = max(FixedPoint(0), share_reserves - z_val)
+    curve_fee_rate = price_discount * curve_fee
+    curve_fee_amount_in_shares = amount_in_shares * curve_fee_rate
+    gov_fee_amount_in_shares = curve_fee_amount_in_shares * gov_fee
+    # applying fee means you get LESS shares out for the same amount of bonds IN
+    amount_to_user_in_shares = amount_in_shares - curve_fee_amount_in_shares
+    return amount_to_user_in_shares, curve_fee_amount_in_shares, gov_fee_amount_in_shares
+
+
+# TODO: switch over to using function in the SDK
+def calc_reserves_to_hit_target_rate(
+    target_rate: FixedPoint, interface: HyperdriveInterface
+) -> tuple[FixedPoint, FixedPoint]:
+    """Calculate the bonds and shares needed to hit the target fixed rate.
+
+    Arguments
+    ---------
+    target_rate : FixedPoint
+        The target rate the pool will have after the calculated change in bonds and shares.
+    interface : HyperdriveInterface
+        The Hyperdrive API interface object.
+
+    Returns
+    -------
+    tuple[FixedPoint, FixedPoint] containing:
+        total_shares_needed : FixedPoint
+            Total amount of shares needed to be added into the pool to hit the target rate.
+        total_bonds_needed : FixedPoint
+            Total amount of bonds needed to be added into the pool to hit the target rate.
+    """
+    # variables
+    predicted_rate = FixedPoint(0)
+    pool_config = interface.pool_config.copy()
+    pool_info = interface.pool_info.copy()
+
+    variable_rate = interface.variable_rate
+    one_block_return = (FixedPoint(1) + variable_rate) ** (
+        FixedPoint(12) / FixedPoint(365 * 24 * 60 * 60)
+    )  # attempt to see if this improves accuracy, it's 1e-8 difference
+
+    iteration = 0
+    start_time = time.time()
+    total_shares_needed = FixedPoint(0)
+    total_bonds_needed = FixedPoint(0)
+    # pylint: disable=logging-fstring-interpolation
+    logging.info(f"Targetting {float(target_rate):.2%} from {float(interface.fixed_rate):.2%}")
+    while float(abs(predicted_rate - target_rate)) > TOLERANCE:
+        iteration += 1
+        target_bonds = calc_bond_reserves(
+            pool_info["shareReserves"],
+            pool_info["shareAdjustment"],
+            pool_config["initialSharePrice"],
+            target_rate,
+            pool_config["positionDuration"],
+            pool_config["invTimeStretch"],
+        )
+        bonds_needed = (target_bonds - pool_info["bondReserves"]) / FixedPoint(2)
```

Comment:
can you add a comment here explaining the division by 2, per our discussion elsewhere.

Reply:
yeah this confused me for a sec, i assume this is part of the newtonian method?